Main Menu Redesign, Part 5

Noordfrees opened the pull request:
This branch is a bit smaller than its predecessors.
- Refactor relationship between LaunchSPG and MapSelect to make 'Back' navigation more intuitive
- Add a navigation bar to the top of the main menu windows to make the stacking more transparent
- The world loading and map generation loadscreen from the Random Match setup is now confined to the window
- Fix LaunchSPG scrolling with 16-player maps
- Fix #4576

Noordfrees committed:
Add navigation bar

Noordfrees committed:
Refactor LaunchSPG/MapSelect relation :)

Noordfrees committed:
Restrict random game progress screen to window

Noordfrees committed:
5 files were automatically formatted.

Noordfrees committed:
fix clang-tidy

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into main-menu-redesign-5

Noordfrees committed:
Fix LaunchSPG scrolling with 16-player maps

Noordfrees committed:
Fix crash when loading invalid game

Noordfrees committed:
3 files were automatically formatted.

Noordfrees committed:
Merge branch 'master' into main-menu-redesign-5

klaus-halfmann reviewed (changes requested):
Please see my inline commnents. 

I got the following warningns from the compiler:
../src/ui_fsmenu/launch_mpg.h:68:7: warning: 'clicked_select_map' overrides a member function but is not marked 'override' [-Winconsistent-missing-override]
        void clicked_select_map();
             ^
../src/ui_fsmenu/launch_game.h:52:15: note: overridden virtual function is here
        virtual void clicked_select_map() = 0;
                     ^
In file included from ../src/ui_fsmenu/mapselect.cc:32:
../src/ui_fsmenu/launch_mpg.h:69:7: warning: 'clicked_select_savegame' overrides a member function but is not marked 'override' [-Winconsistent-missing-override]
        void clicked_select_savegame();
             ^
../src/ui_fsmenu/launch_game.h:53:15: note: overridden virtual function is here
        virtual void clicked_select_savegame() = 0;

These must be fixed.

klaus-halfmann commented:
I am not sure if this "empty" Launch Game Screen still confused a novice user? 
Choosing an intial map  cold still be a default action. 
Other opinions?

Noordfrees committed:
Address review

klaus-halfmann reviewed (changes requested):
I now get
../src/ui_fsmenu/launch_spg.h:43:2: warning: '~LaunchSPG' overrides a destructor but is not marked 'override' [-Winconsistent-missing-destructor-override]
        ~LaunchSPG() = default;

So you better recreate them.

Unrealated (probably just introduced in master)
../src/network/network_lan_promotion.cc:68:4: warning: unknown command tag name [-Wdocumentation-unknown-command]
 * \internal In an ideal world, we would use the same code with boost asio for all three operating
   ^~~~~~~~~
1 warning generated.

Noordfrees committed:
Add missing 'override'

Noordfrees commented:
Fixed.
The `-Wdocumentation-unknown-command` one is in master and has been around for some time.

---

> I am not sure if this "empty" Launch Game Screen still confused a novice user?
Choosing an intial map cold still be a default action.
Other opinions?

I think this would be undesired for two reasons:
- When you want to use a savegame, you will have to click Back before being able to select one, so this would be an extra click then.
- The current LaunchSPG screen behaves like this. Pressing N in the main menu opens two screens, the SPG screen and another MapSelect screen on top of it. This results in confusing behaviour when pressing Back.
    This PR changes the LaunchSPG's behaviour to be more intuitive. Changing the LaunchMPG as proposed would introduce the same inconsistency again.

In an earlier PR I had replaced the Choose Map & Choose Savegame buttons with a dropdown, and automatically opened that dropdown and moved the mouse to the first entry (Choose Map) for convenience, but it was remove again because the automatic mouse movement felt unexpected. And the current screen already says "Please select a map or saved game." so IMHO it is clear to the user what he needs to do.

klaus-halfmann reviewed (approved):
OK for me now.

Playing some Random game now. 
I cannot place my HQ, but I will check this in master again, later.

Noordfrees commented:
Thanks for the review :) 

> Playing some Random game now.
I cannot place my HQ, but I will check this in master again, later.

Probably #1628?
